### PR TITLE
Mock build requires analysis

### LINKED ISSUE #175: implement package_builder.build_rpm method
this is fragment of it's the previous implementation. refactor it. 
```python
p = Popen( ["mock", "-r", distro, srpm_path], stdout=PIPE, stderr=STDOUT)
line = ""
logs = ""
while p.poll() is None:
    line = str(p.stdout.readline())
    if "INFO: Results and/or logs in:" in line:
        logs = line
result = str(self._get_last_word(logs).replace("\\n'", ""))
if not "Finish: run" in line:
    return self._parseError(result)
```

### LINKED ISSUE #176: return errors from package_builder.build_rpm
the method should return list of error strings in case of failure. depends on #175

### PR DESCRIPTION
**Pull request by cibo94**

### CODE REVIEW

File: rpg/package_builder.py [hunk 1]
```diff
@@ -29,7 +18,36 @@ def build_srpm(spec_file, tarball, output_file):
...
+    @staticmethod
+    def build_rpm(srpm, distro, arch):
+
+        def check_output(proc):
+            while proc.poll() is None:
+                line = proc.stdout.readline().decode("utf-8")
+                print(line)
+                if re.search(r"[eE][rR][rR][oO][rR]|" +
+                             r"[eE][xX][cC][eE][pP][tT][iI][oO][nN]",
+                             line):
```
Comment (jsilhan) [inline rpg/package_builder.py]: why not add re.IGNORECASE flag?

**Comment by jsilhan** [comment]
Nice, can you add some test for `build_rpm` that actually passes, please? There could be no requirement in the srpm, just test it created arch specific rpm.

File: rpg/package_builder.py (hunk 1 above)
Comment (cibo94) [inline rpg/package_builder.py]: because python doesn't have support of partial ignorecase - it just doesn't work as we need to

**Comment by cibo94** [comment]
We can't use mock, because mock can't be run from mock - tests will fail on this. I tried this and mock throw an error that user password is wrong on travis so I deleted that test.

File: rpg/package_builder.py (hunk 1 above)
Comment (jsilhan) [inline rpg/package_builder.py]: ok

**Comment by jsilhan** [comment]
@xsuchy , can you answer please, if there is any way to run mock build inside another mock build?
otherwise we would need to do https://github.com/rh-lab-q/rpg/issues/211

**Comment by xsuchy** [comment]
@jsilhan While it is not supported (and especially tested) it works for me. Especially with --new-chroot (which is still experimental feature) it is reduced to question "can container be run inside of container", which again works for me.

**Comment by jsilhan** [comment]
Thanks for response, Mirek.
@cibo94 , please, make it work (https://github.com/rh-lab-q/rpg/issues/212). Btw your commit broke the CI.

File: rpg/package_builder.py [hunk 2]
```diff
@@ -2,9 +2,77 @@
...
+                            _file = match.group(1)
+                        else:
+                            _file = "/usr/bin/" + match.group(1)
+                    break
+            Command("rm -rf " + str(test_package)).execute()
+            if _file == "":
+                break
+            self.base.spec.required_files.append(_file)
+            self.base.spec.build_required_files.append(_file)
+        Command("rm -rf " + str(test_package) +
```
Comment (jsilhan) [inline rpg/package_builder.py]: this code is redundant with line 52. You could do some `cleanup` subfunction.

File: rpg/package_builder.py [hunk 3]
```diff
@@ -2,9 +2,77 @@
...
 class PackageBuilder(object):
+    regex = re.compile(r"[eE][rR][rR][oO][rR]|" +
+                       r"[eE][xX][cC][eE][pP][tT][iI][oO][nN]|" +
+                       r"[cC][oO][mM][mM][aA][nN][dD] [nN][oO][tT] " +
+                       r"[fF][oO][uU][nN][dD]")
+
+    def __init__(self, base):
+        self.base = base
+
+    def mock_build_req(self, distro, arch, temp_dir):
```
Comment (jsilhan) [inline rpg/package_builder.py]: It should be called `build_rpm_recover` and should not build `test_build.src.rpm` but regular srpm.

File: rpg/__init__.py [hunk 4]
```diff
@@ -29,7 +29,7 @@ def __init__(self):
         self._project_builder = ProjectBuilder()
         self.spec = Spec()
         self.sack = None
-        self._package_builder = PackageBuilder()
+        self._package_builder = PackageBuilder(self)
```
Comment (jsilhan) [inline rpg/__init__.py]: don;t pass there Base. it's bad design to have circular dependencies. add `spec` param to `build_rpm_recover` if you need to.

Comment (jsilhan) [inline rpg/__init__.py]: `temp_dir` arg with the default path set could be in constructor and not `temp_dir` arg in `build_rpm` and `build_rpm_recover`

File: rpg/package_builder.py (hunk 3 above)
Comment (cibo94) [inline rpg/package_builder.py]: and what's the difference?

File: rpg/plugins/misc/files_to_pkgs.py [hunk 5]
```diff
@@ -18,7 +18,10 @@ def _resolve(files, query):
...
             logging.info("Resolving Requires")
-            spec.Requires = spec.Requires.union(set(_resolve(spec.required_files, _query)))
+            spec.Requires = spec.Requires.union(
+                set(_resolve(spec.required_files, _query)))
             logging.info("Resolving BuildRequires")
-            spec.BuildRequires = spec.BuildRequires.union(set(_resolve(spec.build_required_files,
-                                                 _query)))
+            spec.BuildRequires = spec.BuildRequires.union(
+                set(_resolve(spec.build_required_files, _query)))
+            spec.required_files = []
```
Comment (jsilhan) [inline rpg/plugins/misc/files_to_pkgs.py]: this is a set now

File: rpg/package_builder.py [hunk 6]
```diff
@@ -2,12 +2,92 @@
...
+                    else:
+                        if "/" in match.group(1):
+                            _file = match.group(1)
+                        else:
+                            _file = "/usr/bin/" + match.group(1)
+                    break
+            cleanup([test_package])
+            if _file == "":
+                break
+            spec.required_files.append(_file)
```
Comment (jsilhan) [inline rpg/package_builder.py]: it's a set now

File: tests/unit/test_plugins.py [hunk 7]
```diff
@@ -80,7 +81,8 @@ def test_find_files(self):
                     ('/libs/__pycache__/', r'%exclude', None),
                     ('/archives/__pycache__/', r'%exclude', None),
                     ('/__pycache__/', r'%exclude', None),
-                    ('/srpm/__pycache__/', '%exclude', None)]
+                    ('/srpm/__pycache__/', '%exclude', None),
+                    ('/mock_project/__pycache__/', '%exclude', None)]
```
Comment (jsilhan) [inline tests/unit/test_plugins.py]: python plugin should not append these at first place. It should only add exclude when *.py file exists in parent dir (could be fixed later).

**Pull request merged**